kymatio/scattering2d/backend/numpy_backend.py

```diff
@@ -0,0 +1,178 @@
+# Authors: Edouard Oyallon, Sergey Zagoruyko
+
+import numpy as np
+from collections import namedtuple
+
+BACKEND_NAME = 'numpy'
+
+class Pad(object):
+    def __init__(self, pad_size, input_size, pre_pad=False):
+        """
+            Padding which allows to simultaneously pad in a reflection fashion
+            and map to complex.
+            Parameters
+            ----------
+            pad_size : list of 4 integers
+                size of padding to apply.
+            input_size : list of 2 integers
+                size of the original signal
+            pre_pad : boolean
+                if set to true, then there is no padding, one simply adds the imaginarty part.
+        """
+        self.pre_pad = pre_pad
+        self.pad_size = pad_size
+
+    def __call__(self, x):
+        if self.pre_pad:
+            return x
+        else:
+            np_pad = ((self.pad_size[0], self.pad_size[1]), (self.pad_size[2], self.pad_size[3]))
+            batch_shape = x.shape[:-2]
+            signal_shape = x.shape[-2:]
+            x = x.reshape((-1, 1) + signal_shape)
+            output = np.pad(x, ((0,0), (0,0), np_pad[0], np_pad[1]), mode='reflect')
+            output = output.reshape(batch_shape + output.shape[-2:])
+            return output
+
+
+def unpad(in_):
+    """
+        Slices the input tensor at indices between 1::-1
+        Parameters
+        ----------
+        in_ : tensor_like
+            input tensor
+        Returns
+        -------
+        in_[..., 1:-1, 1:-1]
+    """
+    return np.expand_dims(in_[..., 1:-1, 1:-1],-3)
+
+class SubsampleFourier(object):
+    """
+        Subsampling of a 2D image performed in the Fourier domain
+        Subsampling in the spatial domain amounts to periodization
+        in the Fourier domain, hence the formula.
+        Parameters
+        ----------
+        x : tensor_like
+            input tensor with at least 5 dimensions, the last being the real
+             and imaginary parts.
+            Ideally, the last dimension should be a power of 2 to avoid errors.
+        k : int
+            integer such that x is subsampled by 2**k along the spatial variables.
+        Returns
+        -------
+        res : tensor_like
+            tensor such that its fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, x, k):
+        batch_shape = x.shape[:-2]
+        signal_shape = x.shape[-2:]
+        
+        x = x.reshape((-1,) + signal_shape)
+        y = x.reshape(-1, k, x.shape[1] // k, k, x.shape[2] // k)
```

Comment:
Torch backend uses -3 instead of -2, and view instead of reshape.

Reply:
Again, this should be handled by the frontend.